### PR TITLE
[NEON] Speedup float16 convert

### PR DESCRIPTION
**Pull request by malfet**
By using `vcvt_f16_f32` and back

According to [benchmark_convert.py](https://github.com/malfet/llm_experiments/commit/d3279637ca925a2e28afea9f8d28080e4ecc23de) this makes float32 to float16 tensor conversion roughly 3 times faster: time to convert 4096x4096 float32 tensor drops from  5.23 msec to 1.66 msec on M2 Pro

Test plan: run `vector_test_all_types` + CI

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h [hunk 1]
```diff
@@ -1049,7 +1048,35 @@ inline Vectorized<type> convert_float_##name(const Vectorized<float>& a, const V
   return Vectorized<type>::loadu(arr2); \
 }
 CONVERT_NON_VECTORIZED_INIT(BFloat16, bfloat16);
+#if defined(__aarch64__) && !defined(C10_MOBILE) && !defined(__CUDACC__)
+inline std::tuple<Vectorized<float>, Vectorized<float>> convert_half_float(const Vectorized<Half>& a) {
+  constexpr int64_t K = Vectorized<Half>::size();
+  __at_align__ float16_t arr2[K];
+  a.store(arr2);
```
Comment (kimishpatel) [inline aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:1055]: Why do we need to do this? Can you use this https://fburl.com/code/t13r938d?

File: aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h [hunk 2]
```diff
@@ -1049,7 +1048,35 @@ inline Vectorized<type> convert_float_##name(const Vectorized<float>& a, const V
...
+  float32x4_t x1 = vcvt_f32_f16(vget_low_f16(x));
+  float32x4_t x2 = vcvt_f32_f16(vget_high_f16(x));
+  float16x8_t y = vld1q_f16(arr2 + Vectorized<float>::size());
+  float32x4_t y1 = vcvt_f32_f16(vget_low_f16(y));
+  float32x4_t y2 = vcvt_f32_f16(vget_high_f16(y));
+  return { Vectorized<float>(x1, x2), Vectorized<float>(y1, y2) };
+}
+inline Vectorized<Half> convert_float_half(const Vectorized<float>& a, const Vectorized<float>& b) {
+  constexpr int64_t K = Vectorized<Half>::size();
+  __at_align__ float16_t arr2[K];
```
Comment (kimishpatel) [inline aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:1066]: This seems to assume K==16. Can you assert?

**Review by kimishpatel — APPROVED** [review]
Left a couple of comments. See if those make sense

File: aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h (hunk 1 above)
Comment (malfet) [reply aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:1055]: Good point (I though that store is not guaranteed to be aligned, but it is, so let me make a change here)

**Pull request closed without merging**